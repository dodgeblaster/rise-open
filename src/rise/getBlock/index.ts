import { RiseCommandInput, RiseBlock } from '../types'
const qlmerge = require('@graphql-tools/merge')
const fs = require('fs')
const path = require('path')

/**
 * Get Code By Type from Block
 * TODO: this needs to be typed corectly
 * @param block
 * @param type
 */
const getCodeByTypeFromBlock = (block: any, type: string) => {
    if (!block.code || !block.code[type]) {
        return {}
    }

    const codeBlocks = Object.keys(block.code[type] || [])
    const formatIntoRiseDeclaration = (x: any) => {
        return typeof x === 'function' ? 'FUNCTION' : x
    }

    return codeBlocks.reduce((acc, k) => {
        const target = block.code[type][k]
        let toAdd = {}

        if (Array.isArray(target)) {
            toAdd = target.reduce((acc: any, x: any) => {
                acc.push(formatIntoRiseDeclaration(x))
                return acc
            }, [])
        } else {
            toAdd = formatIntoRiseDeclaration(target)
        }

        return {
            ...acc,
            [k]: toAdd
        }
    }, {})
}

const getRiseBlock = async (pathString: string, isModule: boolean) => {
    let x
    try {
        x = require(pathString)
    } catch (e) {
        throw new Error('Your project does not have a rise.js file')
    }

    if (Object.keys(x).length === 0) {
        throw new Error(pathString + ' does not have anything defined')
    }

    if (!x.api || typeof x.api !== 'string') {
        try {
            x.api = fs.readFileSync(
                path.join(pathString, '/schema.graphql'),
                'utf8'
            )
        } catch (e) {
            throw new Error(`${pathString} does have a valid GraphQL schema`)
        }
    }

    if (!x.code || Object.keys(x.code).length === 0) {
        throw new Error(module + ' does not have code defined')
    }

    if (!isModule && !x.config) {
        throw new Error(
            'The rise.js file at the root of your project must have config defined'
        )
    }

    if (!isModule && !x.config.name) {
        throw new Error(
            'The rise.js file at the root of your project must have a name inside config'
        )
    }

    x.code.Query = getCodeByTypeFromBlock(x, 'Query')
    x.code.Mutation = getCodeByTypeFromBlock(x, 'Mutation')
    return x
}

const getModuleDirectories = (source: string) => {
    try {
        return fs
            .readdirSync(source, { withFileTypes: true })
            .filter((dirent: any) => dirent.isDirectory())
            .map((dirent: any) => dirent.name)
    } catch (e) {
        return []
    }
}

async function getRootRiseBlock(pathPrefix: string) {
    const block: any = await getRiseBlock(pathPrefix + '/rise.js', false)

    const moduleFolders = getModuleDirectories(pathPrefix + '/modules')

    for (const module of moduleFolders) {
        const path = pathPrefix + '/modules/' + module + '/rise.js'
        const nestedBlock = await getRiseBlock(path, true)

        // Merge code into root rise code
        block.code.Query = {
            ...block.code.Query,
            ...nestedBlock.code.Query
        }
        block.code.Mutation = {
            ...block.code.Mutation,
            ...nestedBlock.code.Mutation
        }

        // Merge schema into root rise schema
        if (block.api) {
            const ast = qlmerge.mergeTypeDefs([block.api, nestedBlock.api])
            block.api = qlmerge.printWithComments(ast)
        } else {
            block.api = nestedBlock.api
        }
    }

    const noValues = (x: any) => Object.keys(x).length === 0
    if (noValues(block.code.Query) && noValues(block.code.Mutation)) {
        throw new Error('You must have at least 1 Query or Mutation defined')
    }

    if (block.api.trim().length === 0) {
        throw new Error('You must have GraphQL types defined')
    }

    return block
}

async function getBlock(
    input: RiseCommandInput,
    pathPrefix: string
): Promise<RiseBlock> {
    let block = await getRootRiseBlock(pathPrefix)

    const name = block.config.name
    if (!name) {
        throw new Error('Rise App must have a name defined')
    }

    const profile = input.profile || block.config.profile || 'default'
    const region = input.region || block.config.region || 'us-east-1'
    const stage = input.stage || block.config.stage || 'dev'
    const auth = block.config.auth || false
    const env = block.config.env || {}
    const bucketName = 'rise-lambdadeployments'
    const bucketFile = `rise-${name}-${stage}.zip`

    return {
        api: block.api,
        code: block.code,
        config: {
            name: name,
            auth: auth,
            profile: profile,
            region: region,
            stage: stage,
            env: env,
            s3BucketFile: bucketName,
            s3BucketName: bucketFile
        }
    }
}

export default getBlock
