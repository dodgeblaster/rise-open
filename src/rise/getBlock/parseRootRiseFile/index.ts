const fs = require('fs')
const { parse, print } = require('recast')

function getExportedRiseBlock(ast: any) {
    return ast.program.body.find((x: any) => {
        return (
            x.type === 'ExpressionStatement' &&
            x.expression.operator === '=' &&
            x.expression.left.type === 'MemberExpression' &&
            x.expression.left.object.name === 'module' &&
            x.expression.left.property.name === 'exports' &&
            x.expression.right.type === 'ObjectExpression'
        )
    })
}

function getRiseConfig(riseBlock: any) {
    return riseBlock.expression.right.properties.find((x: any) => {
        return x.value.type === 'ObjectExpression' && x.key.name === 'config'
    })
}

function getRiseCode(riseBlock: any) {
    return riseBlock.expression.right.properties.find((x: any) => {
        return x.value.type === 'ObjectExpression' && x.key.name === 'code'
    })
}

function getCodeType(code: any, type: string) {
    const section = code?.value?.properties.find((x: any) => {
        return x.value.type === 'ObjectExpression' && x.key.name === type
    })

    return section
}

function getRootRiseFile(pathString: string) {
    const ast = parse(fs.readFileSync(pathString))
    const riseBlock = getExportedRiseBlock(ast)
    // const config = getRiseConfig(riseBlock)
    // const code = getRiseCode(riseBlock)
    //const Query = getCodeType(code, 'Query')

    return print(riseBlock).code
}

export default getRootRiseFile
