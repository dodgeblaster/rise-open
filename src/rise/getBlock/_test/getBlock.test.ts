import getBlock from '../index'

test('can get Block', async () => {
    const projectPath = __dirname + '/exampleProject'

    const result = await getBlock(
        {
            profile: 'example',
            region: 'us-east-2',
            stage: 'dev',
            name: 'blueapp'
        },
        projectPath
    )

    expect(result).toMatchSnapshot()
})
