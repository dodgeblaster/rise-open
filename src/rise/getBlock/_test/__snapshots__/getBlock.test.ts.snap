// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can get Block 1`] = `
Object {
  "api": "type Query {
  note: String
  hat: String
}

schema {
  query: Query
}
",
  "code": Object {
    "Events": Object {},
    "Mutation": Object {},
    "Query": Object {
      "hat": Object {
        "type": "function",
      },
      "note": Object {
        "type": "function",
      },
    },
  },
  "config": Object {
    "additionalUserPool": null,
    "auth": false,
    "env": Object {},
    "events": Array [],
    "name": "hi",
    "profile": "example",
    "region": "us-east-2",
    "s3BucketFile": "rise-hi-dev.zip",
    "s3BucketName": "rise-lambdadeployments",
    "stage": "dev",
  },
}
`;
